### ISSUE TITLE
Strict Zigbee Cluster Library (R8) and Zigbee PRO 2017 (R22) Protocol Specification compatibility for out-of-the-box with the ZHA integration in Home Assistant?

### ISSUE THREAD

**Issue by Hedda** [stanvn/zigbee-plant-sensor#3]
See ZHA docs which indirectly explain that ZHA integration more or less follows the standard Zigbee specifications by the letter:

https://www.home-assistant.io/integrations/zha/#knowing-which-devices-are-supported

https://www.home-assistant.io/integrations/zha/#zha-exception-and-deviation-handling

Summary; a device should only need a “quirk” in the ZHA Device Handlers library if don’t use standard Zigbee clusters and attributes, so receommend strictly follow Zigbee Cluster Library (R8) and Zigbee PRO 2017 (R22) Protocol Specification for out-of-the-box compatibility with zigpy library and the ZHA integration in Home Assistant:

https://github.com/zigpy/zha-device-handlers/blob/dev/README.md#what-are-these-specifications

**_What are these specifications_**

* [Zigbee PRO 2017 (R22) Protocol Specification](https://zigbeealliance.org/wp-content/uploads/2019/11/docs-05-3474-21-0csg-zigbee-specification.pdf)
* [Zigbee Cluster Library (R8)](https://zigbeealliance.org/wp-content/uploads/2021/10/07-5123-08-Zigbee-Cluster-Library.pdf)
* [Zigbee Base Device Behavior Specification (V1.0)](https://zigbeealliance.org/wp-content/uploads/zip/zigbee-base-device-behavior-bdb-v1-0.zip)
* [Zigbee Lighting & Occupancy Device Specification (V1.0)](https://zigbeealliance.org/wp-content/uploads/2019/11/docs-15-0014-05-0plo-Lighting-OccupancyDevice-Specification-V1.0.pdf)
* [Zigbee Primer](https://docs.smartthings.com/en/latest/device-type-developers-guide/zigbee-primer.html)

https://github.com/zigpy/zha-device-handlers/

https://github.com/zigpy/zha-device-handlers/blob/dev/README.md

Also see discussion in this somewhat related issue:

https://github.com/zigpy/zha-device-handlers/issues/1171

Still, if the device gets complexed or advanced with more sensors then might also be nice to have a "Blueprint" for Home Assistant:

https://www.home-assistant.io/blog/2020/12/13/release-202012/#blueprints

https://www.home-assistant.io/docs/blueprint/

https://www.home-assistant.io/docs/automation/using_blueprints/

https://community.home-assistant.io/c/blueprints-exchange/53?page=1

PS: Blueprints are universal could also be used for Zigbee2MQTT and Phoscon/deCONZ so not limited to the ZHA integration.

**Comment by oleo65** [comment]
Hey,

this looks very promising. I just went down the road the other day to get started with the nRF SDK to finally port the bluetooth firmware to Zigbee.

Therefore I am happy to pitch in here and to see this going forward. 

I have a Home Assistant / ZHA setup and am willing to try to get this working. 😃

**Comment by stanvn** [comment]
I have not tested it with ZHA yet. Based on all the information from @Hedda at least the temperature, humidity and illuminance should work. For the soil moisture I created a new cluster that is know by Zigbee2MQTT but not part of the standard [Zigbee Cluster Library ](https://zigbeealliance.org/wp-content/uploads/2019/12/07-5123-06-zigbee-cluster-library-specification.pdf). 

If anyone can test it with ZHA it would be much help. I'll also test it as soon as possible.

**Comment by oleo65** [comment]
I might be able to do a test tonight and will report back with more info.

**Comment by Hedda** [comment]
ZHA/zigpy support Water Content Measurement clusters for Soil Moisture and Leaf Wetness, see these for reference:

https://github.com/home-assistant/core/pull/59300

https://github.com/zigpy/zigpy/pull/844

**Comment by oleo65** [comment]
Ok, I took the code *as is* and compiled and flashed the firmware to a pristine b-parasite.

The discovery in Home Assistant via ZHA integration worked out of the box. 🎉 

It also recognized all available properties, although some values seem to be off or misinterpreted.

![image](https://user-images.githubusercontent.com/4728549/199596156-abbf0884-5363-4acb-86a0-ed9f483669b1.png)

Temperature und Humidity seem to be fine. For Illuminance, a value is reported but there seems to be some sort of error while interpreting the value (it was quite dark < 50 lux)

~Soil moisture and~ battery is not working right now. For battery it seems to me as a first guess that ZHA is maybe expecting the value as battery percentage instead of voltage.

Edit: Soil moisture is giving a value. It was just lying on the table and showing 0 %. I simply expected to show it a (very) low number but not zero and assumed that it was not working correctly. 😇 

Bottom line: This is a really good starting point and very promising. I hope I have some more time the next days for improvements. 😃

**Comment by Hedda** [comment]
If need help tweeking values/parameters for ZHA or discuss what and how to show all attributes correctly as device entities in Home Assistant by default in ZHA then suggest create a new [Device Support Request] issues in the ZHA Device Handlers repository to open a discussion with the developers there -> https://github.com/zigpy/zha-device-handlers/issues

Note! Make first sure that the firmware has its own unique Zigbee device signature as otherwise it will not be possible to diffrience this device from others and make a custom ”quirk” in ZHA Device Handlers for it if needed.

**Comment by stanvn** [comment]
Awesome @oleo65 thanks for testing. 

The battery is a known issue (also with Zigbee2MQTT). It is possible to manually readout the values, but somehow it is not reporting the values on its own. 

For Illuminance, it looks to me that it is not converting the raw transmitted value back to a lux value. According to the Zigbee Cluster Library raw transmitted value should be $10000 \times \log_{10}{\text{Illuminance}} + 1$. I'll look into this as soon as I have a ZHA setup running.

**Comment by oleo65** [comment]
I added some of my findings into two separate pull requests:

* Battery level reporting via ZHA #4 
* Illuminance reportings fixed #5

**Comment by Hedda** [comment]
@stanvn See that rbaron and oleo65 are working on Zigbee firmware sample for firmware + related improvements for upstream:

* Introduce a Zigbee firmware sample  -> https://github.com/rbaron/b-parasite/pull/81
* Zigbee improvements -> https://github.com/rbaron/b-parasite/pull/84

rbaron is btw also working on a new hardware revision with buttons that should make Zigbee joining and device resetting easier:

https://github.com/rbaron/b-parasite/tree/hardware-v2

**Comment by Hedda** [comment]
> @stanvn See that rbaron and oleo65 are working on Zigbee firmware sample for firmware + related improvements for upstream:
> 
> * Introduce a Zigbee firmware sample  -> [Introduce an experimental/educactional/exploratory Zigbee firmware sample rbaron/b-parasite#81](https://github.com/rbaron/b-parasite/pull/81)
> * Zigbee improvements -> [Zigbee improvements rbaron/b-parasite#84](https://github.com/rbaron/b-parasite/pull/84)

@stanvn  FYI, [rbaron](https://github.com/rbaron) has now merged that Zigbee sample to b-parasite upstream repository and he hasd tested it with Home Assistant's ZHA integration (which works without modifications) as well as with Zigbee2MQTT (which requires zigbee-herdsman converter `b-parasite.js` from his [PR](https://github.com/rbaron/b-parasite/pull/81) that needs to be added as per [Z2M instructions to support new devices](https://www.zigbee2mqtt.io/advanced/support-new-devices/01_support_new_devices.html)):

https://github.com/rbaron/b-parasite/tree/main/code/nrf-connect/samples/zigbee

**Comment by stanvn** [comment]
@Hedda thanks for your updates. I think this issue can be closed now. Feel free to reopen it if something comes up.
